src/libraries/Native/Unix/System.Security.Cryptography.Native/android_security.c

```diff
@@ -0,0 +1,196 @@
+        mdName = (jstring)(*jniEnv)->NewStringUTF(jniEnv, "MD5");
+    else if (type == CryptoNative_EvpSha1())
+        mdName = (jstring)(*jniEnv)->NewStringUTF(jniEnv, "SHA-1");
+    else if (type == CryptoNative_EvpSha256())
+        mdName = (jstring)(*jniEnv)->NewStringUTF(jniEnv, "SHA-256");
+    else if (type == CryptoNative_EvpSha384())
+        mdName = (jstring)(*jniEnv)->NewStringUTF(jniEnv, "SHA-384");
+    else if (type == CryptoNative_EvpSha512())
+        mdName = (jstring)(*jniEnv)->NewStringUTF(jniEnv, "SHA-512");
+    else
+        return NULL;
+
+    return (*jniEnv)->CallStaticObjectMethod(jniEnv, mdClass, mdGetInstanceMethodId, mdName);
+}
+
+int32_t CryptoNative_EvpDigestOneShot(intptr_t type, void* source, int32_t sourceSize, uint8_t* md, uint32_t* mdSize)
+{
+    if (!type || !md || !mdSize || sourceSize < 0)
+        return 0;
+
+    JNIEnv* jniEnv = GetJniEnv();
+
+    // MessageDigest md = MessageDigest.getInstance("...");
+    // hashed = md.digest(src);
+
+    jclass mdClass = (*jniEnv)->FindClass(jniEnv, "java/security/MessageDigest");
+    jmethodID mdDigestMethodId = (*jniEnv)->GetMethodID(jniEnv, mdClass, "digest", "([B)[B");
+
+    // TODO: for one-shot digest()s it makes sense to cache instance (convert to a gref)
+    jobject mdObj = GetMessageDigestInstance(jniEnv, type);
+    if (!mdObj)
+        return 0;
+
+    jbyteArray bytes = (*jniEnv)->NewByteArray(jniEnv, sourceSize);
+    (*jniEnv)->SetByteArrayRegion(jniEnv, bytes, 0, sourceSize, (jbyte*) source);
+    jbyteArray hashedBytes = (jbyteArray)(*jniEnv)->CallObjectMethod(jniEnv, mdObj, mdDigestMethodId, bytes);
+
+    jsize hashedBytesLen = (*jniEnv)->GetArrayLength(jniEnv, hashedBytes);
+    (*jniEnv)->GetByteArrayRegion(jniEnv, hashedBytes, 0, hashedBytesLen, (jbyte*) md);
+    *mdSize = (uint32_t)hashedBytesLen;
+
+    // TODO: should we delete all locals by hands here?
+
+    return SUCCESS;
+}
+
+void* CryptoNative_EvpMdCtxCreate(intptr_t type)
+{
+    JNIEnv* jniEnv = GetJniEnv();
+    jobject md = ToGRef(jniEnv, GetMessageDigestInstance(jniEnv, type));
+    return (void*)md;
+}
+
+int32_t CryptoNative_EvpDigestReset(void* ctx, intptr_t type)
+{
+    (void)type;
+
+    JNIEnv* jniEnv = GetJniEnv();
+    jobject mdObj = (jobject)ctx;
+    jclass mdClass = (*jniEnv)->FindClass(jniEnv, "java/security/MessageDigest");
+    jmethodID mdResetMethodId = (*jniEnv)->GetMethodID(jniEnv, mdClass, "reset", "()V");
+    (*jniEnv)->CallVoidMethod(jniEnv, mdObj, mdResetMethodId);
+    return SUCCESS;
+}
+
+int32_t CryptoNative_EvpDigestUpdate(void* ctx, void* d, int32_t cnt)
+{
+    JNIEnv* jniEnv = GetJniEnv();
+    jobject mdObj = (jobject)ctx;
+    jclass mdClass = (*jniEnv)->FindClass(jniEnv, "java/security/MessageDigest");
+    jmethodID mdUpdateMethodId = (*jniEnv)->GetMethodID(jniEnv, mdClass, "update", "([B)V");
+
+    jbyteArray bytes = (*jniEnv)->NewByteArray(jniEnv, cnt);
+    (*jniEnv)->SetByteArrayRegion(jniEnv, bytes, 0, cnt, (jbyte*) d);
```

Comment:
As part of error checking functions which return `void`, read the JNI docs, e.g. https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#Set_PrimitiveType_ArrayRegion_routines

Which documents:

> **THROWS**:
> `ArrayIndexOutOfBoundsException`: if one of the indexes in the region is not valid.

How do you know if it threw?  By using [`JNIEnv::ExceptionOccurred()`](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#ExceptionOccurred).